### PR TITLE
Demo textarea

### PR DESCRIPTION
**Pull request by nfour**
An input by which users can input a definition file body directly.

- [x] Visual overlay in `Demo` for a `textarea`
- [x] Allow `Lincoln` to accept alternative to definitionUrl, `definition`, being a body
- [x] Refactor demo state control
- [x] Minor refactor; `specUrl` to `definitionUrl` props for consistancy

### CODE REVIEW

**Comment by nfour** [comment]
@brendo Ready for review

File: src/entries/demo/Demo.js [hunk 1]
```diff
@@ -16,38 +18,77 @@ const hash = window.location.hash
...
+  updateDefinition = (event) => this.setState({ useDefinition: false, definition: event.target.value })
+  toggleDialog = () => this.setState({ showDialog: !this.state.showDialog })
 
-  setUrlInput = (input) => {
-    this.urlInput = input
+  renderInputDefinition = () => {
+    this.setState({ useDefinition: true, definitionUrl: '', showDialog: false })
   }
 
   render () {
```
Comment (nfour) [inline src/entries/demo/Demo.js:36]: I recognize this is getting bloated - not sure if worth refactoring yet.

File: src/containers/Base/Base.js [hunk 2]
```diff
@@ -27,65 +26,95 @@ export default class Base extends React.PureComponent {
...
+  updateDefinition = async (props) => {
+    const { parserType } = this.state
+    const {
+      definitionUrl, navSort, validate
+    } = props
+
+    let { definition } = props
+
+    if (definitionUrl && !definition) {
+      if (!definitionUrl) { return }
```
Comment (brendo) [inline src/containers/Base/Base.js]: I don't think this condition can ever be true given the line above?

File: src/entries/demo/Demo.js (hunk 1 above)
Comment (brendo) [reply src/entries/demo/Demo.js:36]: Nah, let's wait until the demo is pulled away... need to try and keep velocity up while we get features in.

File: src/containers/Base/Base.js [hunk 3]
```diff
@@ -27,65 +26,95 @@ export default class Base extends React.PureComponent {
...
+        definition,
+        parsedDefinition,
+        parserType
+      })
+    } catch (error) {
+      return this.setState({ loading: false, error })
     }
   }
 
+  fetchDefinition = async ({ definitionUrl, validate, parserType = this.state.parserType }) => {
```
Comment (brendo) [inline src/containers/Base/Base.js:94]: I like the split out, thanks.

File: src/lib/definitions.js [hunk 4]
```diff
@@ -31,6 +22,10 @@ export function validateDefinition (definition, parserType) {
 }
 
 export async function parseDefinition ({ definition, parserType, navSort }) {
+  if (typeof definition === 'string') {
+    definition = YAML.safeLoad(definition)
```
Comment (brendo) [inline src/lib/definitions.js:26]: Question for another day, reckon the parser should be responsible for this? Or just be dumb and always expect a string?

**Review by brendo — COMMENTED** [review]
Haven't checked this out locally yet, but it LGTM.

File: src/containers/Base/Base.js [hunk 5]
```diff
@@ -138,7 +168,7 @@ const Definition = ({ definition, definitionUrl, hash }) =>
       <h3>Render your Open API definition by adding the CORS-enabled URL above.</h3>
       <p>You can also set this with the <code>?url</code> query parameter.</p>
```
Comment (brendo) [inline src/containers/Base/Base.js]: Guess this copy can be updated too, as there's a new way to do it ;)

**Review by brendo — CHANGES_REQUESTED** [review]
Minor changes to default state text and to the README to document the `definition` property required.

File: src/containers/Base/Base.js (hunk 2 above)
Comment (nfour) [reply src/containers/Base/Base.js]: Ha, this is why we have reviews

File: src/containers/Base/Base.js (hunk 5 above)
Comment (nfour) [reply src/containers/Base/Base.js]: :ok_hand:

File: src/lib/definitions.js (hunk 4 above)
Comment (nfour) [reply src/lib/definitions.js:26]: No reason why you can't give it pre-parsed structures, for efficiency. The type check is trivial, and the types are very deterministic.

**Comment by nfour** [comment]
@brendo feedback implimented

**Pull request closed without merging**